src/tpm2-tss-engine-tcti.c

```diff
@@ -0,0 +1,426 @@
+//**********************************************************************;
+// Copyright (c) 2018, General Electric Company.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice,
+// this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimer in the documentation
+// and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <dlfcn.h>
+
+#include <tss2/tss2_tcti.h>
+#include <openssl/err.h>
+
+#include "tpm2-tss-engine-err.h"
+
+#define TMP2TSS_TCTI_NAMEFORMAT "libtss2-tcti-%s.so"
+#define TPM2TSS_TCTI_ENVVAR "TPM2TSSENGINE_TCTI"
+
+/*  Define an explicit dl_handle type to reduce confusion */
+typedef void* dl_handle_t;
+
+static char*
+tcti_expand_dlname (const char *shortname)
+{
+    char *expanddlname;
+    /*  determine required buffer length and allocate it*/
+    int size = snprintf (NULL,
+                         0,
+                         TMP2TSS_TCTI_NAMEFORMAT,
+                         shortname);
+    if (size <= 0) {
+        ERR(tcti_expand_dlname, TPM2TSS_R_GENERAL_FAILURE);
+        expanddlname = NULL;
+    } else {
+        expanddlname = (char*)malloc((size_t)(size+1));
```

Comment:
I think you can even allocate the dlname on stack, i.e.
```
char[size+1] expanddlname;
```

Reply:
I thought about it and did some quick tests, and I think it's not a good idea to allocate `expanddlname` on the stack, because the TCTI option string is under user control with no length validation.

Specifically, an environment variable can be set to an arbitrarily large string. I ran the following code:
```
    #! /bin/bash

    export LARGE_ENVVAR="0123456789abcdef"
    for i in `seq 1 20`;
    do
        export LARGE_ENVVAR="${LARGE_ENVVAR}${LARGE_ENVVAR}"
    done
    echo ${LARGE_ENVVAR} > /tmp/large-file
    unset LARGE_ENVVAR
    ls -l /tmp/large-file
    rm -f /tmp/large-file
```
and the size of `/tmp/large-file` was 16MB!!

Also, there is a limit to how far we can grow the stack, and how large a local variable can be. I compiled and ran the following code:
```
    #include <stdio.h>

    void main(void)
    {
        unsigned long i;
        for (i=0; i<64; i++) {
            unsigned char stack_array[1<<i];
            fprintf(stderr, "    stack size: 2^%lu\n", i);
        }
    }
```
and it segfaulted after printing `stack size: 2^22`.

On the other hand, when we allocate space on the heap, there is necessarily size validation. If there is not enough space, `malloc` or `OPENSSL_malloc` will fail and we check for that.

I do agree with using `OPENSSL_malloc` rather than `malloc`.